sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java

```diff
@@ -306,14 +323,21 @@ public void populateDisplayData(DisplayData.Builder builder) {
 
     /**
      * Writes to file(s) with the given output prefix. See {@link FileSystems} for information on
-     * supported file systems.
-     *
-     * <p>The name of the output files will be determined by the {@link FilenamePolicy} used.
+     * supported file systems. This prefix is used by the {@link DefaultFilenamePolicy}
+     * to generate filenames.
      *
      * <p>By default, a {@link DefaultFilenamePolicy} will build output filenames using the
      * specified prefix, a shard name template (see {@link #withShardNameTemplate(String)}, and
      * a common suffix (if supplied using {@link #withSuffix(String)}). This default can be
      * overridden using {@link #withFilenamePolicy(FilenamePolicy)}.
+     *
+     * <p>This default policy can be overridden using {@link #withFilenamePolicy(FilenamePolicy)},
+     * in which case {@link #withShardNameTemplate(String)} and {@link #withSuffix(String)} should
+     * not be set. Custom filename policies do not automatically see this prefix - you should
+     * explicitly pass the prefix into your {@link FilenamePolicy} object if you need this.
```

Comment:
right now to() is used to pick up the temp directory if a custom filename policy is used.

Another possibility - get rid of withFilenamePolicy and replace it with a to(FilenamePolicy) override. Incompatible, but the FilenamePolicy stuff is all marked as experimental.